### PR TITLE
KAFKA-7250: fix transform function in scala DSL to accept TranformerS…

### PR DESCRIPTION
**Pull request by mdziemianko**
…upplier

Restructuring scala DSL transform function to accept TransformerSupplier instead of a single
instance of Transformer that was shared across tasks.

Updated scaladoc.

Added a unit test to ensure created topology corresponds to equivalent java definition.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala [hunk 1]
```diff
@@ -24,13 +24,12 @@ import java.util.regex.Pattern
 import org.scalatest.junit.JUnitSuite
 import org.junit.Assert._
 import org.junit._
-
```
Comment (guozhangwang) [inline streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala]: Is it intentional to remove the newlines?

**Review by guozhangwang — COMMENTED** [review]
LGTM! Just one minor question.

**Comment by mdziemianko** [comment]
@guozhangwang the formatting change was a mistake. I reverted it.

**Comment by guozhangwang** [comment]
Cherry-picked to 2.0 as well.

**Pull request merged**